server/src/main/java/timely/store/iterators/TimeSeriesGroupingIterator.java

```diff
@@ -0,0 +1,309 @@
+package timely.store.iterators;
+
+import java.io.IOException;
+import java.io.Serializable;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Set;
+import java.util.TreeMap;
+
+import org.apache.accumulo.core.data.ByteSequence;
+import org.apache.accumulo.core.data.Key;
+import org.apache.accumulo.core.data.Range;
+import org.apache.accumulo.core.data.Value;
+import org.apache.accumulo.core.iterators.IteratorEnvironment;
+import org.apache.accumulo.core.iterators.SortedKeyValueIterator;
+import org.apache.accumulo.core.iterators.WrappingIterator;
+import org.apache.accumulo.core.util.Pair;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import timely.adapter.accumulo.MetricAdapter;
+import timely.model.Metric;
+
+/**
+ * Iterator that groups time series so that a filter can be applied to the time
+ * series. This iterator will ignore time series that do not have enough data
+ * points at the start of the query.
+ *
+ */
+public class TimeSeriesGroupingIterator extends WrappingIterator {
+
+    /**
+     * Object representing a single time series and its computed value
+     *
+     */
+    private static class TimeSeries extends TreeMap<Key, Double> implements Serializable {
+
+        private static final long serialVersionUID = 1L;
+        private int targetSize;
+        private transient TimeSeriesGroupingIterator iter;
+        private Double answer;
+
+        public TimeSeries(TimeSeriesGroupingIterator iter, int size) {
+            this.iter = iter;
+            this.targetSize = size;
+        }
+
+        @Override
+        public Double put(Key key, Double value) {
+            return this.put(key, value, true);
+        }
+
+        public Double put(Key key, Double value, boolean doCompute) {
+            if (this.targetSize == super.size()) {
```

Comment:
suggestion....if the targetSize is N, then we only need to keep N-1 values along with the computed value.